source/Missions/Services/Agents/Packets/MovementHandler.cs

```diff
@@ -81,50 +67,91 @@ public void Dispose()
         {
             _packetManager.RemovePacketHandler(this);
             _messageBroker.Unsubscribe<PeerDisconnected>(Handle_PeerDisconnect);
-            m_AgentPollingCancelToken.Cancel();
-            m_AgentPollingTask.Wait();
+            _messageBroker.Unsubscribe<ActionDataChanged>(Handle_ActionDataChanged);
+            _messageBroker.Unsubscribe<LookDirectionChanged>(Handle_LookDirectionChanged);
+            _messageBroker.Unsubscribe<MountDataChanged>(Handle_MountDataChanged);
+            _messageBroker.Unsubscribe<MovementInputVectorChanged>(Handle_MovementInputVectorChanged);
+            _senderTimer?.Dispose();
         }
 
         public PacketType PacketType => PacketType.Movement;
 
-        Mission CurrentMission
+        private void Handle_ActionDataChanged(MessagePayload<ActionDataChanged> payload)
+        {
+            var delta = GetDelta(payload.What);
+
+            delta.CalculateMovement(payload.What);
+        }
+
+        private void Handle_LookDirectionChanged(MessagePayload<LookDirectionChanged> payload)
+        {
+            var delta = GetDelta(payload.What);
+
+            delta.CalculateMovement(payload.What);
+        }    
+
+        private void Handle_MountDataChanged(MessagePayload<MountDataChanged> payload)
         {
-            get
+            var delta = GetDelta(payload.What);
+
+            delta.CalculateMovement(payload.What);
+        }
+
+        private void Handle_MovementInputVectorChanged(MessagePayload<MovementInputVectorChanged> payload)
+        {
+            var delta = GetDelta(payload.What);
+
+            delta.CalculateMovement(payload.What);
+        }
+
+        private AgentMovement GetDelta(IMovementEvent payload)
+        {
+            if (!_agentRegistry.TryGetAgentId(payload.Agent, out var payloadGuid))
             {
-                Mission current = null;
-                GameLoopRunner.RunOnMainThread(() =>
-                {
-                    current = Mission.Current;
-                }, true);
-                return current;
+                Logger.Error("No {agent} found", nameof(Agent));
+            }
+
+            if (_agentMovementDeltas.TryGetValue(payloadGuid, out var delta))
```

Comment:
I could be wrong but if we have an older delta stored here we would use that one instead of the newer one with this logic. It should be we always use the newer one.

Reply:
but at this point we are only getting the deltas that will be used as a base for further calculations.
They will then all be enumerated in PopAllDeltas() and the Dictionary cleared. But I'm also not that happy with a dictionary which basically turned into a queue. Maybe I could change this, too.